### PR TITLE
Fixed performance of Min/Max in SortedSet.

### PR DESCRIPTION
**Pull request by mjp41**
The existing implementation of Min/Max calls InOrderTreeWalk,
which perform an allocation of a stack for traversing the tree.
But when you are finding the smallest or largest element in a
tree there is no need to perform this allocation.  This is an
optimised version that directly finds the smallest or largest
element.

This is an important use case for using a sorted set as a priority 
queue.  In microbenchmarking, with the following code I 
found a 30-40% speed up:
```csharp
        for(int j=0; j < 10; j++) {
            var s = new SortedSet<object>();
            for(int n = 0; n < 100000; n++) {
                s.Add(n);
            }

            int i = 0;
            while(s.Min != null) 
            {
                var o = s.Min;
                s.Remove(o);
                i++;
                if(i%2 == 0) {
                    s.Add((int)o * 2);
                }
            }
        }
```

### CODE REVIEW

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 1]
```diff
@@ -1845,19 +1845,37 @@ public T Min
...
             {
-                T ret = default(T);
-                InOrderTreeWalk(delegate (SortedSet<T>.Node n) { ret = n.Item; return false; });
-                return ret;
+                if (_root == null)
+                {
+                    return default(T);
+                }
+
+                var current = _root;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Nit: please replace ```var``` with ```Node```. We try to only use ```var``` when the type is obvious from the right-hand side of the assignment.

File: src/System.Collections/src/System/Collections/Generic/SortedSet.cs [hunk 2]
```diff
@@ -1845,19 +1845,37 @@ public T Min
...
             {
-                T ret = default(T);
-                InOrderTreeWalk(delegate (SortedSet<T>.Node n) { ret = n.Item; return false; }, true);
-                return ret;
+                if (_root == null)
+                {
+                    return default(T);
+                }
+
+                var current = _root;
```
Comment (stephentoub) [inline src/System.Collections/src/System/Collections/Generic/SortedSet.cs]: Ditto

**Comment by mjp41** [comment]
Thanks @stephentoub, I have fixed the code review feedback. 

There is a similar problem with SortedDictionary, but it is much harder to fix as I think it requires an API change. I'll raise an issue for this, as the choices need discussing.

**Comment by stephentoub** [comment]
@dotnet-bot Test Innerloop CentOS7.1 Release Build and Test please ("Failed to mkdirs")

**Pull request merged**